### ISSUE TITLE
Using sakoechibe window is slower than nowindow

### ISSUE THREAD

**Issue by vancromy** [DynamicTimeWarping/dtw-python#18]
* Python port of R's Comprehensive Dynamic Time Warp algorithm package version: 1.1.10
* Python version: Python 3.9.7 (default, Nov  4 2021, 19:19:42) 
* Operating System: macOS Monterrey 12.1

### Description

Describe what you were trying to get done. 
I'm trying to correlate to 1-D series with the DTW algorithm and wanted to control the window size used during warping. Applying the sakoechiba window constraint seems to massively slow down the algorithm execution. Is this intended?

My query and reference curves are of length ~12000. 

Tell us what happened, what went wrong, and what you expected to happen.
I expected using windowing would in fact speed up execution rather than slow it down?

### What I Did

```
start = time()
alignment = dtw(
    query,
    reference,
    dist_method="euclidean",
    window_type="sakoechiba",
    window_args={"window_size": 20},
    keep_internals=True,
)
duration = time() - start
print(f"{duration=}")

output: 194.4200530052185

start = time()
alignment = dtw(
    query,
    reference,
    dist_method="euclidean",
    window_type=None,
    keep_internals=True,
)
duration = time() - start
print(f"{duration=}")

output: 19.95380139350891
```

**Comment by tonigi** [comment]
Uhm... not intended. The only reason I can think of is that the calculation is based on the use of IEEE NaN, which slows down some processors. Alternatively, you may be at the edge of available RAM. Does the behaviour occur, say, with timeseries 1/4 as long?

**Comment by vancromy** [comment]
Yes behaviour also happens with a slice of the timeseries (I tested 1/4 as suggested just now):
```
duration=11.969528198242188 (with sakoechiba window of size 20)
duration=0.2997860908508301 (no window)
```

RAM doesn't seem to be an issue either. I have 32GB ram and my monitor shows it only went up to 18GB.

**Comment by tonigi** [comment]
Then I’m afraid it’s related to the hardware NaN implementation. What processor do you have?

It would need more tests. Rewriting is possible but needs to be done in the R version as well, and thoroughly tested, for which i don’t have a timeline.

**Comment by tonigi** [comment]
Perhaps try a different installation system (e.g. conda, vs. pip, vs. pip compiled from source). Compiling from source may enable you to pass different optimization / floating point switches. But I wouldn't know from the top of my head which ones.

**Comment by vancromy** [comment]
Just to update the thred. I tried a poetry install, a build from source and a pip install git+repo:url but get same performance across all of these cases. 

If anyone else is able to reproduce this, please update the thread.

**Comment by tonigi** [comment]
Hmm, upon testing, the NaN thing is negligible. 

The more likely cause of slowdown is possibly due to  the way the `sakoeChibaWindow` and similar functions are called https://github.com/DynamicTimeWarping/dtw-python/blob/6c21db2c946c794b06ed3c958b55ad8033b266f2/dtw/_globalCostMatrix.py#L16 .   

The code is that way for generality, which I wouldn't want to sacrifice. I'll keep an eye about alternatives though.

**Comment by tonigi** [comment]
Please try 571727cd1d2d04fb89bfe6c35c0f24a503f8f837 i.e. v1.1.11

**Comment by vancromy** [comment]
Problem fixed! Thanks for the fast response and a great package :) 

Vectorizing the windowing algorithm operations seems to have done the trick!

```
duration=4.609917163848877 (with window)
duration=5.144525766372681 (without)
```
